conftest.py

```diff
@@ -234,3 +237,57 @@ def enable_mlflow_testing():
     with pytest.MonkeyPatch.context() as mp:
         mp.setenv(_MLFLOW_TESTING.name, "TRUE")
         yield
+
+
+@pytest.fixture(scope="session", autouse=True)
+def local_pypi_repo(tmp_path_factory):
+    """
+    A local PyPI repository that serves a wheel for the current MLflow version.
+    """
+    root = tmp_path_factory.mktemp("root")
+    mlflow_dir = root.joinpath("mlflow")
+    mlflow_dir.mkdir()
+    port = get_safe_port()
+    try:
+        repo_root = subprocess.check_output(
+            [
+                "git",
+                "rev-parse",
+                "--show-toplevel",
+            ],
+            text=True,
+        ).strip()
+    except subprocess.CalledProcessError:
+        # Some tests run in a Docker container where git is not installed.
+        # In this case, assume we're in the root of the repo.
+        repo_root = "."
+
+    subprocess.run(
+        [
+            sys.executable,
+            "-m",
+            "pip",
+            "wheel",
+            "--wheel-dir",
+            mlflow_dir,
+            "--no-deps",
+            repo_root,
+        ],
+        check=True,
+    )
```

Comment:
This command is slow (takes a few seconds). pytest runs this when a test session starts.

1. Make this opt-in (= add a flag)
2. Cache the wheel and refresh it when necessary (e.g. refresh daily).

@B-Step62 Let me know if you come with a better approach.

Reply:
I think caching doesn't work as we should  update the mlflow wheel for basically every code change. For the first option, where do you think is appropriate to add that flag?

Another option would be having a small wrapper for pip install/download command, that checks mlflow is in required and in dev version, then spawn pypi server on the fly. Need to be careful to starting building multiple servers tho.

Last resort is go back to local wheel solution just to workaround cloudpickle issue here.... but I kinda like this approach more.

Reply:
> Last resort is just to build local wheel that particularly solves cloudpickle issue.... but I kinda like this local server approach more.

Sounds good. Let's this first :)

Reply:
> For the first option, where do you think is appropriate to add that flag?

We can add the flag in `conftest.py`:

```
# conftest.py
def pytest_addoption(parser):
    parser.addoption(
        "--serve-wheel",
        ...
    )
```

Reply:
> I'm curious what the code looks like :) Is it possible to file a draft?

Sure! Will file a draft PR in minutes.

> We can add the flag in `conftest.py:`

Ah I see, sounds doable. A slight concern about visibility... probably related test cases should catch pip error and show some message about the flag.

Reply:
Another option is opt-out.

Reply:
@B-Step62 The code (for opt-out) looks like this:

```diff
diff --git a/conftest.py b/conftest.py
index 4ecd911aa7..e3792411dc 100644
--- a/conftest.py
+++ b/conftest.py
@@ -42,6 +42,11 @@ def pytest_addoption(parser):
         type=int,
         help="The group of tests to run.",
     )
+    parser.addoption(
+        "--no-serve-wheel",
+        action="store_true",
+        help="Disable serving the wheel for the current MLflow version for faster test execution.",
+    )
 
 
 def pytest_configure(config):
@@ -240,10 +245,14 @@ def enable_mlflow_testing():
 
 
 @pytest.fixture(scope="session", autouse=True)
-def local_pypi_repo(tmp_path_factory):
+def local_pypi_repo(request, tmp_path_factory):
     """
     A local PyPI repository that serves a wheel for the current MLflow version.
     """
+    if request.config.getoption("--no-serve-wheel"):
+        yield  # pytest expects a generator fixture to yield
+        return
+
     root = tmp_path_factory.mktemp("root")
     mlflow_dir = root.joinpath("mlflow")
     mlflow_dir.mkdir()
```

Reply:
This is still a bit inconvenient because you need to use `--no-serve-wheel` but we can avoid `hey, why this test doesn't pass on my machine but does on CI?`.

Reply:
Thanks for the code sample. Hmm opt-out is actually more concerning in terms of visibility; like if it is opt-in, developers see pip error when it is required, ideally with suggestion of `--serve-wheel` option in error message. On the other hand, if it is opt-out, I guess many ppl are suffered from that longer waiting without knowing.

What do you think about this option or https://github.com/mlflow/mlflow/pull/10262?
```
Another option would be having a small wrapper for pip install/download command, that checks mlflow is in required and in dev version (+ in test or not), then spawn pypi server on the fly. Need to be careful to starting building multiple servers tho.
```

Reply:
Can #10262 fix this test failure: https://github.com/mlflow/mlflow/actions/runs/6677805050/job/18148176722#step:10:42?

```
tests/models/test_cli.py::test_predict FAILED | MEM 0.9/6.8 GB | DISK 43.2/83.2 GB [ 24%]

...

2023/10/28 15:47:29 INFO mlflow.utils.virtualenv: Installing dependencies
ERROR: Cannot install -r requirements.txt (line 1) and cloudpickle==3.0.0 because these package versions have conflicting dependencies.
ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

Reply:
Nope, need to add similar fix to `virtualenv.py` ([code](https://github.com/mlflow/mlflow/blob/84c985585fa40db2a70ea97e62d687adc417ee20/mlflow/utils/virtualenv.py#L281)). Current draft is only for showing logic, eventually some handling logic will be extracted to shared utils e.g. parse requirements.txt and exclude mlflow.

This works for now as supposedly these 2 are only usage, but I prefer pypi server approach in terms of scalability, hence it's trade-off between usability (visibility) vs scalability.

Reply:
Opt-in sounds good.

> Another option would be having a small wrapper for pip install/download command, that checks mlflow is in required and in dev version (+ in test or not), then spawn pypi server on the fly. Need to be careful to starting building multiple servers tho.

I think this works but seems more complex than this PR's approach.

Reply:
Totally makes sense, it's not ideal to complicate normal flow for the sake of testing.

For opt-in approach, could you just add message suggesting that flag? Probably [here](https://github.com/B-Step62/mlflow/blob/ca559d2afd9187b165f7cca8f11b3bfa285e4fbe/mlflow/models/wheeled_model.py#L224) and [here](https://github.com/mlflow/mlflow/blob/84c985585fa40db2a70ea97e62d687adc417ee20/mlflow/utils/virtualenv.py#L283) (or I can do as follow-up).

Reply:
A follow-up sounds good :)

Reply:
fyi: Started https://github.com/mlflow/mlflow/pull/10264 (will finalize after come back from personal errand)